ocaml/utils/target_systems.ml

```diff
@@ -0,0 +1,96 @@
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+type architecture =
+  | IA32
+  | X86_64
+  | ARM
+  | AArch64
+  | POWER
+  | Z
+
+let architecture () : architecture =
+  match Config.architecture with
+  | "i386" -> IA32
+  | "amd64" -> X86_64
+  | "arm" -> ARM
+  | "arm64" -> AArch64
+  | "power" -> POWER
+  | "s390x" -> Z
+  | arch -> Misc.fatal_errorf "Unknown architecture `%s'" arch
+
+type derived_system =
+  | Linux
+  | MinGW_32
+  | MinGW_64
+  | Win32
+  | Win64
+  | Cygwin
+  | MacOS_like
+  | FreeBSD
+  | NetBSD
+  | OpenBSD
+  | Generic_BSD
+  | Solaris
+  | Dragonfly
+  | GNU
+  | BeOS
+  | Unknown
+
+let derived_system () : derived_system =
+  match architecture (), Config.model, Config.system with
+  | IA32, _, "linux_aout" -> Linux
+  | IA32, _, "linux_elf" -> Linux
+  | IA32, _, "bsd_aout" -> Generic_BSD
+  | IA32, _, "bsd_elf" -> Generic_BSD
+  | IA32, _, "beos" -> BeOS
+  | IA32, _, "cygwin" -> Cygwin
+  | (X86_64 | IA32), _, "macosx" -> MacOS_like
+  | IA32, _, "gnu" -> GNU
+  | IA32, _, "mingw" -> MinGW_32
+  | IA32, _, "win32" -> Win32
+  | POWER, "ppc64le", "elf" -> Linux
+  | POWER, "ppc64", "elf" -> Linux
+  | POWER, "ppc", "elf" -> Linux
+  | POWER, "ppc", "netbsd" -> NetBSD
+  | POWER, "ppc", "bsd_elf" -> OpenBSD
+  | Z, _, "elf" -> Linux
+  | ARM, _, "linux_eabihf" -> Linux
+  | ARM, _, "linux_eabi" -> Linux
+  | ARM, _, "bsd" -> OpenBSD
+  | X86_64, _, "linux" -> Linux
+  | X86_64, _, "gnu" -> GNU
+  | X86_64, _, "dragonfly" -> Dragonfly
+  | X86_64, _, "freebsd" -> FreeBSD
+  | X86_64, _, "netbsd" -> NetBSD
+  | X86_64, _, "openbsd" -> OpenBSD
+  | X86_64, _, "darwin" -> MacOS_like
```

Comment:
I did not closely follow the `darwin`/`macos` branding,
but it might be safer to also match IA32 and Arm64.

Reply:
Done